plugins/inputs/firehose/firehose.go

```diff
@@ -0,0 +1,413 @@
+	// decode base64-encoded data and return them as a slice of byte slices
+	decodedData := make([][]byte, 0)
+	for _, record := range r.body.Records {
+		data, err := base64.StdEncoding.DecodeString(record.EncodedData)
+		if err != nil {
+			return nil, false
+		}
+		decodedData = append(decodedData, data)
+	}
+	return decodedData, true
+}
+
+func (r *firehoseRequest) sendResponse(res http.ResponseWriter) error {
+	responseBody := struct {
+		RequestID    string `json:"requestId"`
+		Timestamp    int64  `json:"timestamp"`
+		ErrorMessage string `json:"errorMessage,omitempty"`
+	}{
+		RequestID:    r.req.Header.Get(requestIDHeader),
+		Timestamp:    time.Now().Unix(),
+		ErrorMessage: statusCodeToMessage[r.responseStatusCode],
+	}
+	response, err := json.Marshal(responseBody)
+	if err != nil {
+		return err
+	}
+	res.Header().Set("content-type", "application/json")
+	res.WriteHeader(r.responseStatusCode)
+	_, err = res.Write(response)
+	return err
+}
+
+// Firehose is an input plugin that collects external metrics sent via HTTP from AWS Data Firhose
+type Firehose struct {
+	ServiceAddress string          `toml:"service_address"`
+	Paths          []string        `toml:"paths"`
+	PathTag        bool            `toml:"path_tag"`
+	ReadTimeout    config.Duration `toml:"read_timeout"`
+	WriteTimeout   config.Duration `toml:"write_timeout"`
+	AccessKey      string          `toml:"access_key"`
+	ParameterTags  []string        `toml:"parameter_tags"`
+
+	tlsint.ServerConfig
+	tlsConf *tls.Config
+
+	TimeFunc
+	Log telegraf.Logger
+
+	wg    sync.WaitGroup
+	close chan struct{}
+
+	listener net.Listener
+
+	telegraf.Parser
+	acc telegraf.Accumulator
+}
+
+func (*Firehose) SampleConfig() string {
+	return sampleConfig
+}
+
+func (f *Firehose) Gather(_ telegraf.Accumulator) error {
+	return nil
+}
+
+func (f *Firehose) SetParser(parser telegraf.Parser) {
+	f.Parser = parser
+}
+
+// Start starts the http listener service.
+func (f *Firehose) Start(acc telegraf.Accumulator) error {
+	if f.ReadTimeout < config.Duration(time.Second) {
+		f.ReadTimeout = config.Duration(time.Second * 10)
+	}
+	if f.WriteTimeout < config.Duration(time.Second) {
+		f.WriteTimeout = config.Duration(time.Second * 10)
+	}
+
+	f.acc = acc
+
+	server := f.createHTTPServer()
+
+	f.wg.Add(1)
+	go func() {
+		defer f.wg.Done()
+		if err := server.Serve(f.listener); err != nil {
+			if !errors.Is(err, net.ErrClosed) {
+				f.Log.Errorf("Serve failed - %s", err.Error())
+			}
+			close(f.close)
+		}
+	}()
+
+	f.Log.Infof("Listening on %s", f.listener.Addr().String())
+
+	return nil
+}
+
+func (f *Firehose) createHTTPServer() *http.Server {
+	return &http.Server{
+		Addr:         f.ServiceAddress,
+		Handler:      f,
+		ReadTimeout:  time.Duration(f.ReadTimeout),
+		WriteTimeout: time.Duration(f.WriteTimeout),
+		TLSConfig:    f.tlsConf,
+	}
+}
+
+// Stop cleans up all resources
+func (f *Firehose) Stop() {
+	if f.listener != nil {
+		f.listener.Close()
```

Comment:
Tbh, not sure. Again just copied this block from `http_listener_v2` plugin and does work as expected. I haven't yet inspected why this is used instead of `f.server.Shutdown`, but I'll definitely have a look.

Reply:
Please take everything from http_listener_v2 with care as this is a _really_ old plugin. Check everything against current go standards!